### PR TITLE
Fix possible crash with embedded fonts, and when more than 64 nested elements

### PR DESCRIPTION
**Pull request by poire-z**
Will allow closing https://github.com/koreader/koreader/issues/6394 and https://github.com/koreader/koreader/issues/6391 . Details in these issues.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/crengine/358)

### CODE REVIEW

**Comment by poire-z** [comment]
@virxkane : I believe your https://github.com/virxkane/coolreader/commit/5f604b4ddd4292717606e01c5894b588ca2ae973 is an attempt to fix the same issue as my 2nd commit here, right?

**Comment by virxkane** [comment]
@poire-z Yes, it looks like it. I corrected it in the morning, committed it only recently, did not think that you were busy with the same. I will probably revert it.

File: crengine/src/lvtinydom.cpp [hunk 1]
```diff
@@ -4549,7 +4549,12 @@ bool ldomDocument::render( LVRendPageList * pages, LVDocViewCallback * callback,
             printf("CRE: styles re-init needed after load, re-rendering\n");
             // We should clear RenderRectAccessor, that may have been used for
             // caching CSS checks results (i.e. :nth-child(), :last-of-type...)
-            getRootNode()->clearRenderDataRecursive();
+            if ( _rectStorage.hasChunks() ) {
+                // (We would crash in the following if no RenderRectAccessor has
+                // been created, which may happen if we're here not because of
+                // CSS checks but because of embedded fonts.)
+                getRootNode()->clearRenderDataRecursive();
+            }
```
Comment (virxkane) [inline crengine/src/lvtinydom.cpp:4557]: This code looks a bit ugly, you check property of _rectStorage and later not use _rectStorage at all.

Comment (poire-z) [reply crengine/src/lvtinydom.cpp:4557]: Yes, I know - but it avoids checking it each time for each node in clearRenderDataRecursive().
May be I should just rename and wrap it into a ldomDocument method, `ldomDocument->hasRenderData()` ?

Comment (virxkane) [reply crengine/src/lvtinydom.cpp:4557]: It will look more logical.

**Comment by poire-z** [comment]
@virxkane : my fix about `clearRenderDataRecursive()` seems not enough (and I guess your one would also not be enough).
I get the same crash with the book from another issues opened today after the previous fix: https://github.com/koreader/koreader/issues/6413#issuecomment-662483884.

The same thing is happening (embedded fonts), but in the first case that we both solved in different ways, we had, when about to do it, 0 chunk. So, we skipped doing it when 0 chunk.

But in the case of this book, we have 3 chunks when about to do it - and we crash when doing it.
If skipping the clearRenderDataRecursive(), and looking at the number of chunks later, I'd get 4.

If I do this:
```diff
--- a/crengine/src/lvtinydom.cpp
+++ b/crengine/src/lvtinydom.cpp
@@ -15605,10 +15605,12 @@ void ldomNode::clearRenderData()
 void ldomNode::clearRenderDataRecursive()
 {
     ASSERT_NODE_NOT_NULL;
     if ( !isElement() )
         return;
+    lvdomElementFormatRec tmp;
+    getDocument()->_rectStorage.getRendRectData(_handle._dataIndex, &tmp);
     lvdomElementFormatRec rec;
     getDocument()->_rectStorage.setRendRectData(_handle._dataIndex, &rec);
     int cnt = getChildCount();
     for (int i=0; i<cnt; i++) {
         ldomNode * child = getChildNode( i );
```
No crash !

These methods are quite similar in what they test and do:
https://github.com/koreader/crengine/blob/28ebac45f7b1dc196836fbc86d6cba3424f14c6b/crengine/src/lvtinydom.cpp#L2688-L2722

The only subtle difference is one has `<` and the other `<=` in `while ( _chunks.length() <= chunkIndex )`.

I don't understand anything about this chunk & storage stuff.
As you fixed some related stuff recently, may be you have a better idea of what's going on and what might be wrong and how to fix that?

(I haven't tested yet, but may be we can solve this with using `_chunks.clear()` which might need to realloc later what we cleared - but I think we might have a more general issue there that may trigger in other situations.)

**Comment by virxkane** [comment]
I don't understand anything about this chunk & storage stuff too. Just using debugger, asan, etc.
And no, my fix that I already rolled back is a working solution for this problematic file. https://github.com/virxkane/coolreader/commit/5f604b4ddd4292717606e01c5894b588ca2ae973
Probably just guessing :)

**Comment by poire-z** [comment]
> And no, my fix that I already rolled back is a working solution for this problematic file. 

Doesn't work for me with that file:
```
07/22/20-21:38:34 INFO  opening file Issue6413-embedded_font_crash_scrambled.epub
CRE: document loaded, but styles re-init needed (cause: embedded fonts)
07/22/20-21:38:35 DEBUG CreDocument: loading done.
07/22/20-21:38:35 DEBUG Typography lang: not overriding en-US with doc language: en-US
07/22/20-21:38:35 DEBUG CreDocument: rendering document...
CRE: styles re-init needed after load, re-rendering
Thread 1 "luajit" received signal SIGSEGV, Segmentation fault.
ldomDataStorageManager::getChunk (this=this@entry=0x5555561dc820, address=address@entry=196608) at crengine/src/lvtinydom.cpp:2617
2617         if ( chunk->_prevRecent )
(gdb) bt
#0  ldomDataStorageManager::getChunk (this=this@entry=0x5555561dc820, address=address@entry=196608) at crengine/src/lvtinydom.cpp:2617
#1  0x00007fffecadfa92 in ldomDataStorageManager::setRendRectData (this=0x5555561dc820, elemDataIndex=<optimized out>, src=src@entry=0x7fffffffd470)
    at crengine/src/lvtinydom.cpp:2719
#2  0x00007fffecae09e8 in ldomNode::clearRenderDataRecursive (this=0x55555640f120) at crengine/src/lvtinydom.cpp:15611
#3  0x00007fffecae0a24 in ldomNode::clearRenderDataRecursive (this=0x55555640f110) at crengine/src/lvtinydom.cpp:15616
#4  0x00007fffecae0a24 in ldomNode::clearRenderDataRecursive (this=0x555556f295a0) at crengine/src/lvtinydom.cpp:15616
#5  0x00007fffecae0a24 in ldomNode::clearRenderDataRecursive (this=0x555556f29590) at crengine/src/lvtinydom.cpp:15616
#6  0x00007fffecae0c75 in ldomDocument::render (this=this@entry=0x5555561cc6c0, pages=pages@entry=0x555555ec15d8, callback=0x555555e47580, width=width@entry=582, dy=dy@entry=568,
    showCover=showCover@entry=false, y0=0, def_font=..., def_interline_space=100, props=...) at crengine/src/lvtinydom.cpp:4556
#7  0x00007fffecb9461c in LVDocView::Render (this=0x555555ec1530, dx=582, dx@entry=0, dy=568, dy@entry=0, pages=0x555555ec15d8, pages@entry=0x0)
    at crengine/src/../include/crlocks.h:50
#8  0x00007fffece11980 in renderDocument (L=<optimized out>) at cre.cpp:674
```
```diff
--- a/crengine/src/lvtinydom.cpp
+++ b/crengine/src/lvtinydom.cpp
@@ -4550,12 +4550,10 @@ bool ldomDocument::render( LVRendPageList * pages, LVDocViewCallback * callback,
             printf("CRE: styles re-init needed after load, re-rendering\n");
             // We should clear RenderRectAccessor, that may have been used for
             // caching CSS checks results (i.e. :nth-child(), :last-of-type...)
-            if ( hasRenderData() ) {
                 // (We would crash in the following if no RenderRectAccessor has
                 // been created, which may happen if we're here not because of
                 // CSS checks but because of embedded fonts.)
                 getRootNode()->clearRenderDataRecursive();
-            }
         }
         CRLog::info("rendering context is changed - full render required...");
         // Clear LFormattedTextRef cache
@@ -15608,12 +15606,15 @@ void ldomNode::clearRenderDataRecursive()
     if ( !isElement() )
         return;
     lvdomElementFormatRec rec;
-    getDocument()->_rectStorage.setRendRectData(_handle._dataIndex, &rec);
-    int cnt = getChildCount();
-    for (int i=0; i<cnt; i++) {
-        ldomNode * child = getChildNode( i );
-        if ( child->isElement() ) {
-            child->clearRenderDataRecursive();
+    ldomDataStorageManager& rectStorage = getDocument()->_rectStorage;
+    if (rectStorage.getUncompressedSize() > 0) {
+        rectStorage.setRendRectData(_handle._dataIndex, &rec);
+        int cnt = getChildCount();
+        for (int i=0; i<cnt; i++) {
+            ldomNode * child = getChildNode( i );
+            if ( child->isElement() ) {
+                child->clearRenderDataRecursive();
+            }
         }
     }
 }
```

**Comment by virxkane** [comment]
Strange.
Screenshot_20200722_235204

**Comment by poire-z** [comment]
And do you crash with my patch you picked, and without yours ?

**Comment by virxkane** [comment]
> And do you crash with my patch you picked, and without yours ?

Yes.

**Comment by poire-z** [comment]
And on your side, with your patch, what's the result of `rectStorage.getUncompressedSize() > 0` ? Do you do the `setRendRectData` stuff and recursion or do you just have it skipped ?

**Comment by virxkane** [comment]
> I don't understand anything about this chunk & storage stuff.
> As you fixed some related stuff recently, may be you have a better idea of what's going on and what might be wrong and how to fix that?

OK, I'll tell you what I know about these chunk & storage stuff, but I don't think you will learn something new, since you updated quite a lot of things in the lvtinydom.cpp file, and it would be silly to think that you acted without knowing how to work the insides of this file.
I would not like to remember this confusing/spaghetti code, I have been searching for this unfortunate error for a week and I would like to forget this horror https://github.com/virxkane/coolreader/commit/e160b27c21635519ffd6515e7171f88772778419#diff-77f6525e33d2f73-a19deed1d2f73-64d1adeeb8175
It’s also not easy for me to write in English, so prepare to suffer.
Let's start with the ldomNode structure. It contains the anonymous union _data. The _text_ptr field contains a pointer to a text node if the type of this ldomNode is text, i.e. _handle._dataIndex & 0x0F == NT_TEXT and this element is not persistent, i.e. the next node object itself is stored in memory, not its index. Same thing with the _elem_ptr field (NT_ELEMENT type only).
The _pelem_addr and _ptext_addr fields are the element node and text node indices, respectively, if this ldomDone is persistent, i.e. the data of the elements themselves is not in memory, only their indices. ldomNode is converted to persistent before being stored in the cache using the tinyNodeCollection::persist (), ldomNode::persist() functions, respectively, after loading from the cache, all ldomNode elements must be persistent. I throw somehow a crash while loading the cache, it turned out that for some reason several elements were missing this flag, i.e. ldomNode::isPersistent() returns false, so _text_ptr and _elem_ptr were used instead of using the _pelem_addr and _ptext_addr fields. As you already know, I fixed this by adding the appropriate check.
Then, if necessary, the ldomNode can be converted to a non-persistent element using ldomNode::modify().
The collection of these ldomNodes is managed by a complex messy mess of tinyNodeCollection, ldomDataStorageManager, ldomTextStorageChunk, private class CacheFile and I probably forgot to add something else.
ldomDataStorageManager contains the _chunks attribute, a wrapper array for ldomTextStorageChunk elements, and, among other things, it implements transparent saving and loading to / from the cache file in case of insufficient memory (set with the help of * _CACHE_UNPACKED_SPACE * _storageMaxUncompressedSizeFactor, however, you yourself know).
tinyNodeCollection contains two-dimensional arrays ldomNode * _textList [TNC_PART_COUNT] and ldomNode * _elemList [TNC_PART_COUNT]. Or an array of arrays. Inner arrays are allocated as needed, but always of the same size. See tinyNodeCollection::allocTinyNode(). Unused elements are simply empty and occupy memory, but quickly, since memory is allocated as rarely as possible. Accordingly, you do not need to check how many elements are in the array every time, just take which one you need and that's it. At the same time, in tinyNodeCollection::loadNodeData(), data from the cache file is loaded into a separate array, and its size can be smaller than TNC_PART_LEN, and then a new one of a smaller size is written instead of a pointer to the old inner array. And then, as I wrote above, there is a call to an element of the inner array without checking for its size and, as a result, going out of memory.
I can't say more about ldomTextStorageChunk.

**Comment by virxkane** [comment]
> And on your side, with your patch, what's the result of `rectStorage.getUncompressedSize() > 0` ? Do you do the `setRendRectData` stuff and recursion or do you just have it skipped ?

This is another error! Without this patch (just latest commit) asan give this error:
```coolreader/crengine/src/lvtinydom.cpp:2613:21: runtime error: member access within null pointer of type 'struct ldomTextStorageChunk'```

in ldomDataStorageManager::getChunk() chunk is NULL!
Won't you use asan anyway? :)

**Comment by virxkane** [comment]
Vector ldomDataStorageManager::_chunks contains NULL elements, don't know how it better to fix, just add check in ldomDataStorageManager::getChunk() or exclude that situation...

**Comment by virxkane** [comment]
> Do you do the `setRendRectData` stuff and recursion or do you just have it skipped ?

May be just skipped without any cleaning, I not tested.

**Comment by virxkane** [comment]
> The only subtle difference is one has `<` and the other `<=` in `while ( _chunks.length() <= chunkIndex )`

Looks as bug, I think in both cases must be '<=' The same in getStyleData(), setStyleData(). If replace '<' with '<=' in setRectData() you get the same effect as 
```diff
--- a/crengine/src/lvtinydom.cpp
+++ b/crengine/src/lvtinydom.cpp
@@ -15605,10 +15605,12 @@ void ldomNode::clearRenderData()
 void ldomNode::clearRenderDataRecursive()
 {
     ASSERT_NODE_NOT_NULL;
     if ( !isElement() )
         return;
+    lvdomElementFormatRec tmp;
+    getDocument()->_rectStorage.getRendRectData(_handle._dataIndex, &tmp);
     lvdomElementFormatRec rec;
     getDocument()->_rectStorage.setRendRectData(_handle._dataIndex, &rec);
     int cnt = getChildCount();
     for (int i=0; i<cnt; i++) {
         ldomNode * child = getChildNode( i );
```
No crash.

**Comment by poire-z** [comment]
> OK, I'll tell you what I know
> it would be silly to think that you acted without knowing how to work the insides of this file.
> It’s also not easy for me to write in English, so prepare to suffer.

Thanks,
Yes, I touched some bits of all that, and I know just enough to be able to understand and follow what you are explaining (and your english is fine enough, easier in the long explanations when there is some context - sometimes a bit harder with your few words answers without context where I can have doubts about what you mean :)
But I don't get the whole picture of storage, persistence, cache, auto-load/save from/to cache, when things are dropped or allocated, how the cache grows or is re-used... :/
But anyway, except for these rare bugs, it's quite solid - so I don't really need to understand it :)

> Won't you use asan anyway? :)

I just tried again to use it, but somehow, in our context, I can't make it work. I don't know much about the build infrastructure and tools and all that, and I'm not much interested.
But I can use valgrind when needed, that gives me the same kind of info (after I wait 5 minutes for the app to start up and crash, instead of the usual 3 seconds..):
```
Invalid read of size 8
   ldomDataStorageManager::getChunk(unsigned int) (lvtinydom.cpp:2617)
   ldomDataStorageManager::setRendRectData(unsigned int, lvdomElementFormatRec const*) (lvtinydom.cpp:2719)
   ldomNode::clearRenderDataRecursive() (lvtinydom.cpp:15611)
   ldomNode::clearRenderDataRecursive() (lvtinydom.cpp:15616)
   ldomNode::clearRenderDataRecursive() (lvtinydom.cpp:15616)
   ldomNode::clearRenderDataRecursive() (lvtinydom.cpp:15616)
   ldomDocument::render(LVRendPageList*, LVDocViewCallback*, int, int, bool, int, LVProtectedFastRef<LVFont>, int, LVFastRef<CRPropAccessor>) (lvtinydom.cpp:4557)
   LVDocView::Render(int, int, LVRendPageList*) (lvdocview.cpp:2780)
   renderDocument(lua_State*) (cre.cpp:674)
   lj_BC_FUNCC (buildvm_x86.dasc:849)
   lj_ff_coroutine_resume (buildvm_x86.dasc:1733)
   lua_pcall (lj_api.c:1140)
 Address 0x10 is not stack'd, malloc'd or (recently) free'd

Process terminating with default action of signal 11 (SIGSEGV)
 Access not within mapped region at address 0x10
   ldomDataStorageManager::getChunk(unsigned int) (lvtinydom.cpp:2617)
   ldomDataStorageManager::setRendRectData(unsigned int, lvdomElementFormatRec const*) (lvtinydom.cpp:2719)
   ldomNode::clearRenderDataRecursive() (lvtinydom.cpp:15611)
   ldomNode::clearRenderDataRecursive() (lvtinydom.cpp:15616)
   ldomNode::clearRenderDataRecursive() (lvtinydom.cpp:15616)
   ldomNode::clearRenderDataRecursive() (lvtinydom.cpp:15616)
   ldomDocument::render(LVRendPageList*, LVDocViewCallback*, int, int, bool, int, LVProtectedFastRef<LVFont>, int, LVFastRef<CRPropAccessor>) (lvtinydom.cpp:4557)
   LVDocView::Render(int, int, LVRendPageList*) (lvdocview.cpp:2780)
   renderDocument(lua_State*) (cre.cpp:674)
   lj_BC_FUNCC (buildvm_x86.dasc:849)
   lj_ff_coroutine_resume (buildvm_x86.dasc:1733)
   lua_pcall (lj_api.c:1140)
```

> Looks as bug, I think in both cases must be '<=' 

OK. I'll try to understand a bit more of what's done in there later today.

**Comment by virxkane** [comment]
If we analyze the use of _chunks in ldomDataStorageManager, we can see that no null items are added or replaced anywhere. But in chain  setRendRectData() -> ldomDataStorageManager::getChunk() we found null item.
Ok, looks in ldomDataStorageManager::getRendRectData()
```cpp
    while ( _chunks.length() <= chunkIndex ) {
        //if ( _chunks.length()>0 )
        //    _chunks[_chunks.length()-1]->compact();
        _chunks.add( new ldomTextStorageChunk(RECT_DATA_CHUNK_SIZE, this, _chunks.length()) );
        getChunk( (_chunks.length()-1)<<16 );
        compact( 0 );
    }
    ldomTextStorageChunk * chunk = getChunk( chunkIndex<<16 );
```
Here 'while' loop is used to bring the number of _chunk elements to the required, so later call getChunk( chunkIndex<<16 ) is acceptably.
```cpp
ldomTextStorageChunk * ldomDataStorageManager::getChunk( lUInt32 address )
{
    ldomTextStorageChunk * chunk = _chunks[address>>16];
    if ( chunk!=_recentChunk ) {
        if ( chunk->_prevRecent )
            chunk->_prevRecent->_nextRecent = chunk->_nextRecent;
        if ( chunk->_nextRecent )
            chunk->_nextRecent->_prevRecent = chunk->_prevRecent;
        chunk->_prevRecent = NULL;
        if (((chunk->_nextRecent = _recentChunk)))
            _recentChunk->_prevRecent = chunk;
        _recentChunk = chunk;
    }
    chunk->ensureUnpacked();
    return chunk;
}
```
Here in line ```ldomTextStorageChunk * chunk = _chunks[address>>16]``` is used LVPtrVector::operator[](int pos) without any safety checks, so it very easy get access outside of array bounds. This is our case: if in ldomDataStorageManager::setRendRectData() wrote ```while ( _chunks.length() < chunkIndex )``` we have size of _chunks less by one item, but we not get access outside of array bounds since LVPtrVector::add() call LVPtrVector::insert() that call LVPtrVector::reserve() that insert some null items into LVPtrVector::_list. This is how we have null items in _chunks. It's easy to fix replacing ```while ( _chunks.length() < chunkIndex )``` with ```while ( _chunks.length() <= chunkIndex )```.

**Comment by poire-z** [comment]
OK, thanks, I understand.
This indeed solves the issue - and looks alright:
```diff
--- a/crengine/src/lvtinydom.cpp
+++ b/crengine/src/lvtinydom.cpp
@@ -2674,3 +2674,3 @@ void ldomDataStorageManager::setStyleData( lUInt32 elemDataIndex, const ldomNode
     int chunkIndex = index >> STYLE_DATA_CHUNK_ITEMS_SHIFT;
-    while ( _chunks.length() < chunkIndex ) {
+    while ( _chunks.length() <= chunkIndex ) {
         //if ( _chunks.length()>0 )
@@ -2711,3 +2711,3 @@ void ldomDataStorageManager::setRendRectData( lUInt32 elemDataIndex, const lvdom
     int chunkIndex = index >> RECT_DATA_CHUNK_ITEMS_SHIFT;
-    while ( _chunks.length() < chunkIndex ) {
+    while ( _chunks.length() <= chunkIndex ) {
         //if ( _chunks.length()>0 )
```

(The issue was that, when I'm about to `clearRenderDataRecursive()`, I have get/setRenderData a very little number of RenderRectAccessor, but not for all the nodes - so it indeed happens that when I use setRenderData on all nodes in clearRenderDataRecursive(), we find these NULL chunks when the first node that should be in that chunk dit not get thru getRenderData.
In the normal rendering case, we always do a getRenderData first, before doing later setRenderData - so all is fine.)

Btw, I went with `clearRenderDataRecursive()` (which looks expensive, but is not really compared to everything else  - and we do that in some rare case so no worry) because I had no idea with the indirection how to just zero the existing chunks (and what to free() or not in this whole black box).
So, I zero each microchunk (for each node), while I could have just zero'ed each whole chunk if I had more assurance and understanding of this part of the code :/

**Comment by virxkane** [comment]
So, now ldomDataStorageManager::hasChunks(), tinyNodeCollection::hasRenderData() are used for optimization, not bug fixes.

**Comment by poire-z** [comment]
Right :)

**Pull request merged**